Python->C++ exception translation

feltech opened the pull request:
## Description

Closes #947. Establish a strategy for single-sourcing Python<->C++ exception mapping, add a mechanism to translate OpenAssetIO-specific Python exceptions to corresponding C++ exceptions, and augment the pybind11 override macros to do the translation.

This only affects pybind11 "trampoline" functions - i.e. C++ virtual methods that can call out to a Python implementation.

- [x] I have updated the release notes.
- [x] I have updated all relevant user documentation.

## Reviewer Notes

The final commit is optional. A comment in the GitHub issue mentions `createPythonPluginSystemManagerImplementationFactory`, so the final commit was added to be able to use the new functionality there. However, on inspection it didn't actually seem necessary.

Testing every affected class isn't tackled. Instead, the macros are tested via dedicated contrived `cmodule` (`_openassetio_test`) tests. Testing for all classes can be added if it seems worthwhile, but will be a big chunk of code (similar to the GIL tests). Otherwise, we're OK as long as we trust contributors to always use the `OPENASSETIO_`-prefixed pybind11 macros (as detailed in the CODING_STANDARDS).

Only OpenAssetIO-specific exceptions are converted.  It wouldn't be much extra effort to also cover other  exceptions.

foundrytom commented on src/openassetio-core/include/openassetio/errors/exceptions.hpp:
```diff
@@ -107,6 +108,12 @@ struct OPENASSETIO_CORE_EXPORT BatchElementException : OpenAssetIOException {
   BatchElementError error;
 };
 
+/// List of all OpenAssetIO-specific exceptions. Used in Python
+/// exception translation.
```
[Consider][minor] As this is core, maybe mention languages more generically if the intention is to enable more programatic binding generation for any target language?

feltech replied:
Made more generic in b1373437

foundrytom commented on src/openassetio-core/include/openassetio/errors/exceptions.hpp:
```diff
@@ -107,6 +108,12 @@ struct OPENASSETIO_CORE_EXPORT BatchElementException : OpenAssetIOException {
+/// exception translation.
+using AllExceptions =
+    std::tuple<OpenAssetIOException, InputValidationException, ConfigurationException,
+               NotImplementedException, UnhandledException>;
```
[Question] Why isn't `BatchElementException` here?

foundrytom commented on src/openassetio-python/private/include/openassetio/private/python/exceptions.hpp:
```diff
@@ -0,0 +1,362 @@
+  return true;
+}
+/**
+ * Assert that all exceptions in the AllExceptions list are also found
```
[Consider][minor] For a new reader, it might be nice to differentiate that the above are the mechanism by which the last actually does the assertion.

```suggestion
 * Execute compile-time assertions that all exceptions in the AllExceptions list are also found
```

feltech replied:
Updated docstring as suggested in 960a54e4

foundrytom commented on src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp:
```diff
@@ -11,87 +14,376 @@
+void registerPyExceptionClass<ConfigurationException>(const std::string_view pyClassName,
+                                                      const py::module_ &mod,
+                                                      PyObjectByName &registeredExceptions) {
+  const py::object pyBaseClass = registeredExceptions.at("InputValidationException");
```
[question][naive] Thinking out loud, from a very uninformed starting point, could you use `kCppExceptionsAndPyClassNames` to auto-derive `pyBaseClass` from the introspected base of the template so you dont need a specialisations? Might not be worth it.

elliotcmorris replied:
Heh, i noted in my review how I thought resisting the urge to do this was a good call :P.
Either works, but don't let the genericism override the clarity.

feltech replied:
Had some fun in f4993c24 - can drop if you think it's excessive (it only helps remove a single case). It does have the neat side-effect of ensuring that the Python class hierarchy mirrors C++.

foundrytom replied:
> It does have the neat side-effect of ensuring that the Python class hierarchy mirrors C++.

For me thats a useful benefit, as it makes future maintenance easier and less prone to human error. @elliotcmorris what are your thoughts?

elliotcmorris replied:
> It does have the neat side-effect of ensuring that the Python class hierarchy mirrors C++.

Suppose that's fair. It's not really that much more complicated either.

foundrytom commented on src/openassetio-python/tests/cmodule/test_errors.py:
```diff
@@ -66,3 +70,82 @@ def test_when_ConfigurationException_thrown_then_can_be_caught_as_InputValidatio
+    return exception
+
+
+class PyExceptionThrower(_openassetio_test.ExceptionThrower):
```
[Consider] This is really well documented in c++ - might be worth a cursory note here on what this is doing and how.

feltech replied:
Added a docstring in 9103b05c

foundrytom commented on doc/contributing/CODING_STANDARDS.md:
```diff
@@ -232,6 +232,25 @@ a null pointer.
+pybind11 macros.
+
+So instead of using the `PYBIND11_OVERRIDE_*` family of macros, we must
+use `OPENASSETIO_PYBIND11_OVERRIDE_*`, which decorates the pybind11
```
[question] if wonder if its worth a grep test in CI to assert this? like we do for OAIO?

feltech replied:
Yeah good idea. I was thinking we could change the `business-language` check to `disallowed-phrases` and add it there. Could also be useful if/when we do inclusive terminology checks too.

feltech replied:
CI job updated in 18647406

Showing that "OAIO" is still caught: https://github.com/OpenAssetIO/OpenAssetIO/actions/runs/6876912503/job/18703496245?pr=1190

Showing that bad pybind11 macros is caught: https://github.com/OpenAssetIO/OpenAssetIO/actions/runs/6876921789/job/18703536758?pr=1190

The ID of the job has changed and its a required check, so we can either change the ID back or update the project settings.

foundrytom replied:
Awesome sauce. The new name is better so we can just update the branch protection requirement.

foundrytom reviewed (commented):
Only superficial comments from me as others are far better placed to comment on the implementation. But great job on the thoroughness of approach/testing. Minor Q - would you imagine squashing the last two commits into the previous commit-but-one?

foundrytom commented:
> . A comment in the GitHub issue mentions createPythonPluginSystemManagerImplementationFactory, so the final commit was added to be able to use the new functionality there. However, on inspection it didn't actually seem necessary.

I think the intent was that that code could raise OpenAssetIO exceptions from Python, that would end up in C++.

feltech commented:
> > . A comment in the GitHub issue mentions createPythonPluginSystemManagerImplementationFactory, so the final commit was added to be able to use the new functionality there. However, on inspection it didn't actually seem necessary.
> 
> I think the intent was that that code could raise OpenAssetIO exceptions from Python, that would end up in C++.

Yeah that was the thought, but looking at the code, no OpenAssetIO exceptions can be raised there. We could wrap the construction of `PythonPluginSystemManagerImplementationFactory` just in case we modify the Python constructor in future to raise an OpenAssetIO exception?

The `instantiate` method is already covered because it's pybind11 trampoline method. It was just potentially the constructor that's not covered.

feltech commented:
> Minor Q - would you imagine squashing the last two commits into the previous commit-but-one?

Yeah I guess that would make the history a bit easier to grok, assuming we're happy to keep the new shared private header, even though its not currently necessary.

foundrytom commented:
> Yeah that was the thought, but looking at the code, no OpenAssetIO exceptions can be raised there. We could wrap the construction of `PythonPluginSystemManagerImplementationFactory` just in case we modify the Python constructor in future to raise an OpenAssetIO exception?

Ahh yes - maybe we need to worry about `ManagerInterfaceFactoryInterface` then? As that calls into Python?

foundrytom commented:
> Ahh yes - maybe we need to worry about `ManagerInterfaceFactoryInterface` then? As that calls into Python?

Oh and `PagerInterface`? And anything else like that? Ie many of the places we had to touch the GIL?

elliotcmorris commented on src/openassetio-python/cmodule/src/_openassetio.hpp:
```diff
@@ -20,10 +20,33 @@ OPENASSETIO_FWD_DECLARE(log, LoggerInterface)
+/**
+ * Decorate PYBIND11_OVERRIDE_NAME with exception type translation.
+ */
+#define OPENASSETIO_PYBIND11_OVERRIDE_NAME(ret_type, cname, name, fn, ...)                      \
```
[Consider] I might think about defining these in their own place within the cmodule dir, `exception_translation_macros.hpp` or something. They get in the way of the "point" of this file imo.

feltech replied:
Good idea, they are using up a lot of space now. The `_openassetio.hpp` has become a bit of a dumping ground.

feltech replied:
Good call. Put in their own header in a8e8db56 - much nicer and YAGNI.

elliotcmorris commented on src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp:
```diff
@@ -462,3 +462,129 @@ void registerExceptions(const py::module &mod) {
+template <class Exception>
+void throwHybridExceptionIfMatches(const std::string_view expectedPyExcName,
+                                   const py::error_already_set &thrownPyExc,
+                                   const std::string_view thrownPyExcName) {
```
[Consider] could you extract the name from the PyExc here in the function, rather than require an external extraction and argument pass??

feltech replied:
Could do, but it would be repeated multiple times (once for every call) rather than once in the caller and passed in.

elliotcmorris replied:
Isn't that already the case? Looks like `throwHybridExceptionIfMatches` is only called from inside `convertPyExceptionAndThrow`, and the extraction is being done each time that function is being called anyway.

feltech replied:
But `convertPyExceptionAndThrow` is only called once.

elliotcmorris replied:
Resolving this as it's not important, even though i still don't quite follow the argument... probably just text chat fog-of-war.

elliotcmorris reviewed (changes requested):
Really solid engineering imo. I couldn't have done this as well.

elliotcmorris commented on src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp:
```diff
@@ -11,87 +14,376 @@
-  // single string parameter (message). However, we need `index` and
-  // `error` parameters to mirror C++.
+/**
+ * @section topy Conversion from C++ exception to Python exception.
```
Oh no it's not. its To-py, Doh!

elliotcmorris commented on src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp:
```diff
@@ -11,87 +14,376 @@
+  static constexpr std::size_t kSize = sizeof...(Type);
+  using Types = std::tuple<Type...>;
+  const std::array<std::string_view, kSize> ids;
+};
```
Not actually sure how that works, if std::array would reject an initializer list or a vector of innapropriate size, or if there'd be an implicit conversion leaving some empty string views on the end.

feltech replied:
Yeah it fails at compile time if you try to give too few or too many elements to the array.

hartmutbehrens commented on src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp:
```diff
@@ -11,87 +14,376 @@
+ * General case of a simple exception (only takes a single string
+ * constructor argument) with OpenAssetIOException as its base class.
+ *
+ * @tparam Exception Corresponding C++ exception type.
```
`tparam` -> template parameter, so probably fine.

feltech commented:
> > Ahh yes - maybe we need to worry about `ManagerInterfaceFactoryInterface` then? As that calls into Python?
> 
> Oh and `PagerInterface`? And anything else like that? Ie many of the places we had to touch the GIL?

As long as calling out to Python from C++ is done transparently via a C++ base class virtual method (i.e. pybind11 trampoline), then exceptions raised in Python will be translated, due to the new/updated `OPENASSETIO_PYBIND11_OVERRIDE*` macros.

If there are places where we explicitly call a method on a `py::object`, then we'd need to decorate that call to translate any OpenAssetIO exceptions that propagate out of it.  But I don't think we do this anywhere, apart from `createPythonPluginSystemManagerImplementationFactory`?

feltech commented on src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp:
```diff
@@ -11,87 +14,376 @@
+  using Type = decltype(kCppExceptionsAndPyClassNames);
+  /// "Array" of C++ exception classes.
+  template <std::size_t I>
+  using Exceptions = std::tuple_element_t<I, typename Type::Types>;
```
Yeah the idea was to make it "look" like an array when used, especially when using `kClassNames` and `Execptions` together with the same index.

I think `std::get` only retrieves values, as opposed to types. Maybe it has specialisations, will take a look.

feltech replied:
`std::decay_t<decltype(std::get<I>(std::declval<Type::Types>()))>` :laughing: - I think the existing version is a bit neater.

elliotcmorris replied:
Yep, me too :P

feltech commented on src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp:
```diff
@@ -11,87 +14,376 @@
+ * compiler make the attempt and so fail on any violated static_asserts.
+ */
+[[maybe_unused]] constexpr bool kIsClassListOrderValid =
+    assertClassesAreInHierarchyOrder(CppExceptionsAndPyClassNames::kIndices);
```
Oh interesting, odd kinda no-op assert, but getting rid of the pointless variable would be good.

feltech replied:
Simplified as suggested in 13b69491

feltech commented on src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp:
```diff
@@ -11,87 +14,376 @@
+ */
+
+/**
+ * @section asserts Compile-time assertions.
```
I'm not keen on that, since the `exceptionsValidation.hpp` header on its own would have lots of red squiggles in my editor (since it depends on stuff defined "above").

I am keen on finding a way to hide some of this assertion logic though, since I totally agree it causes issues with readability.

elliotcmorris replied:
Yeah that's a fair point.

feltech commented on src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp:
```diff
@@ -11,87 +14,376 @@
+}
+
+/// Mapping of Python class name to Python class object.
+using PyObjectByName = std::map<const std::string_view, const py::object>;
```
I tried `unordered_map`, but `std::string_view` wouldn't work as a key.  I assumed there was no `std::hash` overload.  A quick google implies that's not the case and it should be possible, though, so I'll take another look.

feltech replied:
Updated in a16f9571 - the trick was to get rid of the `const` from the key type.

feltech commented on src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp:
```diff
@@ -11,87 +14,376 @@
+void registerPyExceptionClass<ConfigurationException>(const std::string_view pyClassName,
+                                                      const py::module_ &mod,
+                                                      PyObjectByName &registeredExceptions) {
+  const py::object pyBaseClass = registeredExceptions.at("InputValidationException");
```
Similar to Tom's comment: https://github.com/OpenAssetIO/OpenAssetIO/pull/1190#discussion_r1389127543 - it's a nice idea, will take a look and see if it's straightforward enough.

elliotcmorris replied:
Resolving in favour of the other thread.

foundrytom commented:
> As long as calling out to Python from C++ is done transparently via a C++ base class virtual method (i.e. pybind11 trampoline), then exceptions raised in Python will be translated, due to the new/updated `OPENASSETIO_PYBIND11_OVERRIDE*` macros.

Oh nice, I hadn't realised we were already using our custom macros there, so 👍 Does [this](https://github.com/OpenAssetIO/OpenAssetIO/pull/1190#issuecomment-1805633069) need one thought?

feltech commented:
> Does [this](https://github.com/OpenAssetIO/OpenAssetIO/pull/1190#issuecomment-1805633069) need one thought?

I'm not sure what you mean? Were you meant to link to a different comment?

foundrytom commented:
> > Does [this](https://github.com/OpenAssetIO/OpenAssetIO/pull/1190#issuecomment-1805633069) need one thought?
> 
> I'm not sure what you mean? Were you meant to link to a different comment?

Sorry no idea what went on there, the link was meant to be https://github.com/OpenAssetIO/OpenAssetIO/blob/2198d02b8415785effb779ed76eba78ed6362fb9/src/openassetio-python/cmodule/src/managerApi/EntityReferencePagerInterfaceBinding.cpp#L40

feltech commented:
> Sorry no idea what went on there, the link was meant to be
> 
> https://github.com/OpenAssetIO/OpenAssetIO/blob/2198d02b8415785effb779ed76eba78ed6362fb9/src/openassetio-python/cmodule/src/managerApi/EntityReferencePagerInterfaceBinding.cpp#L40

Still not sure I follow. `EntityReferencePagerInterfaceBinding.cpp` already used/has been updated to use `OPENASSETIO_*` macros.

feltech commented on src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp:
```diff
@@ -11,87 +14,376 @@
+/**
+ * Set the current Python exception in this thread.
+ *
+ * Overload to handle more complex case of BatchElementException.
```
With the change in https://github.com/OpenAssetIO/OpenAssetIO/pull/1190#discussion_r1389127543 there are only two special cases remaining (`OpenAssetIOException` and `BatchElementException`), and handling them all in the same method could remove a little duplication.

Will have to sketch it out and see if how neat it looks.

feltech replied:
`if constexpr`ed a bunch in 8de8f03a.  I'm two minds about `registerPyExceptionClass` - it only saves one line of duplication, at the expense of a biig function (albeit mostly comments), but it does at least keep it all together. What do you think?

elliotcmorris replied:
I like the smaller ones above, I think that's an improvement. But I see what you mean about `registerPyExceptionClass`. I do think the if/else fallthrough logic is easier to understand than specialisation dispatch though, despite being a bit less elegant feeling.

foundrytom commented:
> > Sorry no idea what went on there, the link was meant to be
> > https://github.com/OpenAssetIO/OpenAssetIO/blob/2198d02b8415785effb779ed76eba78ed6362fb9/src/openassetio-python/cmodule/src/managerApi/EntityReferencePagerInterfaceBinding.cpp#L40
> 
> Still not sure I follow. `EntityReferencePagerInterfaceBinding.cpp` already used/has been updated to use `OPENASSETIO_*` macros.

Sorry I had somehow slipped back to browsing `main` not your tree.

feltech commented:
I've squashed down the commits, in case anyone wants to check the commit messages/breakdown makes sense.

I've also added a fixup 7ec0b72d to move the static assertions to their own .cpp file, since I figured they were obscuring the public interface of the header, and pessimised build times by re-running the static assertions in every translation unit that included it.

foundrytom reviewed (approved):
Great idea @feltech - like it.
